pa4/checker-pa4.ml

```diff
@@ -1212,121 +1241,101 @@ let main () = begin
          * methods that it declares and override.*)
         let get_parents_methods_tbl class_names super_class_list =
-            let methods_map = Hashtbl.create 255 in
-            List.iter (fun class_name ->
-                    let super_class_option = List.find_opt (fun ((_,cname),_,_)-> class_name = cname) super_class_list in
-                    if not (Option.none = super_class_option) then begin
-                            let ((_, _),_,features) = Option.get super_class_option in
-                            (* add methods to hashtbl starting from the ultimate class that override method *)
-                            List.iter (fun current_method ->
-                                    match current_method with
-                                    | Attribute _ -> ()
-                                    | Method ((_,method_name), formals, method_type, exp) -> begin
-                                            let cur = Hashtbl.find_opt methods_map method_name in
-                                            match cur with 
-                                            | None -> Hashtbl.replace methods_map method_name (formals, class_name, exp);
-                                            | _ -> () 
-                                    end
-                            ) features ;
-                    end;
-            ) class_names ;
-            methods_map
+                (*let methods_map_ = !methods_map in*)
+                let methods_map_ = Hashtbl.create 255 in
+                ordered_method_names := [];
+                List.iter (fun class_name ->
+                        let super_class_option = List.find_opt (fun ((_,cname),_,_)-> class_name = cname) super_class_list in
+                        if not (Option.none = super_class_option) then begin
+                                let ((_, super_class_name),_,features) = Option.get super_class_option in
+                                (* add methods to hashtbl starting from the ultimate class that override method *)
+                                List.iter (fun current_method ->
+                                        match current_method with
+                                        | Attribute _ -> ()
+                                        | Method ((_,method_name), formals, method_type, exp) -> begin
+                                                if not(List.mem method_name !ordered_method_names) then
+                                                        ordered_method_names := !ordered_method_names @ [method_name];
+                                                let _, method_type_val = method_type in
+                                                Hashtbl.replace methods_map_ method_name (formals, class_name, exp);
+                                        end
+                                ) features ;
+                        end;
+                ) class_names ;
+                methods_map := methods_map_;
```

Comment:
Why are we storing the methods map outside of of the method if we're returning it right after this?